File: exercism.io/ruby/bob/bob.rb
class Bob
  def hey(text)
    letters_only = text.scan(/[A-Za-z]/)
    letters_up = letters_only.map(&:upcase)

    case
    when text.match(/^\s+$/) || text.empty?
      "Fine. Be that way!"
    when letters_only == letters_up && !letters_only.empty?
      "Whoa, chill out!"
    when text[-1] == "?"
      "Sure."
    else
      "Whatever."
    end
  end
end

File: exercism.io/elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
        empty?(input) -> "Fine. Be that way!"
        question?(input) -> "Sure."
        uppercase?(input) -> "Whoa, chill out!"
        true -> "Whatever."
    end
  end

  def uppercase?(str) do
    letters_only = Regex.scan(~r{[^0-9,.]}, str) |> to_string
    !empty?(letters_only) && String.upcase(letters_only) == letters_only
  end

  def question?(str) do
    String.at(str, -1) == "?"
  end

  def empty?(str) do
    to_charlist(str) |> Enum.empty?
      || String.match?(str, ~r{^\s+$})
  end
end

